### PR TITLE
Remove several instances of panic on corrupt enum.

### PR DESCRIPTION
**Pull request by tommyp1ckles**
Replacing several instances of PANIC_DUE_TO_CORRUPT_ENUM.

Part of #19229

Signed-off-by: tommyp1ckles <tomhadlaw@gmail.com>

Commit Message:
Additional Description:
Risk Level:  low
Testing:
Docs Changes:
Release Notes:

### CODE REVIEW

**Comment by tommyp1ckles** [comment]
Not sure what's going on with the ci: ```$TEST_TMPDIR defined: output root default is '/build/tmp' and max_idle_secs default is '15'.
tee: /dev/stderr: Permission denied```

File: source/common/upstream/load_balancer_impl.h [hunk 1]
```diff
@@ -336,24 +336,29 @@ class ZoneAwareLoadBalancerBase : public LoadBalancerBase {
 
   HostSet& localHostSet() const { return *local_priority_set_->hostSetsPerPriority()[0]; }
 
-  static HostsSource::SourceType localitySourceType(HostAvailability host_availability) {
+  static absl::optional<HostsSource::SourceType>
+  localitySourceType(HostAvailability host_availability) {
     switch (host_availability) {
     case HostAvailability::Healthy:
-      return HostsSource::SourceType::LocalityHealthyHosts;
+      return absl::optional<HostsSource::SourceType>(HostsSource::SourceType::LocalityHealthyHosts);
```
Comment (lizan) [inline source/common/upstream/load_balancer_impl.h:343]: use absl::make_optional (throughout)

**Review by lizan — COMMENTED** [review]
Not sure why this needed API review though. @alyssawilk PTAL?

**Comment by alyssawilk** [comment]
first pass had https://github.com/envoyproxy/envoy/pull/19695/commits/59190ec6e85debd7a20938bd2ad83a401e46131e
I think you can just to standard review.

**Comment by tommyp1ckles** [comment]
CI is failing due to lower test coverage. I'm working on writing test cases that take care of that now.

**Comment by tommyp1ckles** [comment]
ci script fails due to low coverage, struggling to find a good way to test those code paths.

**Comment by tommyp1ckles** [comment]
Seems like I'm in of a bit of a catch-22 with the format, check-format.fix makes changes that allow me to pass the push hook but those changes seem to fail the ci build.

File: test/per_file_coverage.sh [hunk 2]
```diff
@@ -28,7 +28,7 @@ declare -a KNOWN_LOW_COVERAGE=(
 "source/common/tcp:94.6"
 "source/common/thread:0.0" # Death tests don't report LCOV
 "source/common/tracing:96.1"
-"source/common/upstream:96.0"
+"source/common/upstream:95.8"
```
Comment (alyssawilk) [inline test/per_file_coverage.sh:31]: have you tried testing these by static casting an out of bounds number and using EXPECT_ENVOY_BUG?

Comment (tommyp1ckles) [reply test/per_file_coverage.sh:31]: No, I was fixated on the now missing coverage lines in the case that a none-option is returned. But I'll try that and see if that increases the coverage enough to not require this change 👍 .

**Comment by KBaichoo** [comment]
Seems like you were reviewing this @alyssawilk . PTAL when you can

**Review by alyssawilk — APPROVED** [review]
Nice clean up, thanks!

**Pull request merged**